proxy/tracing_test.go

```diff
@@ -70,61 +72,69 @@ func TestTracingFromWire(t *testing.T) {
 	}
 }
 
-func TestTracingRoot(t *testing.T) {
-	traceContent := fmt.Sprintf("%x", md5.New().Sum([]byte(time.Now().String())))
+func TestTracingIngressSpan(t *testing.T) {
 	s := startTestServer(nil, 0, func(r *http.Request) {
-		th, ok := r.Header[traceHeader]
-		if !ok {
-			t.Errorf("missing %s request header", traceHeader)
-		} else {
-			if th[0] != traceContent {
-				t.Errorf("wrong X-Trace-Header content: %s", th[0])
-			}
+		p := &mocktracer.TextMapPropagator{}
+		_, err := p.Extract(ot.HTTPHeadersCarrier(r.Header))
+		if err != nil {
+			t.Error(err)
 		}
 	})
 	defer s.Close()
 
-	u, _ := url.ParseRequestURI("https://www.example.org/hello")
-	r := &http.Request{
-		URL:    u,
-		Method: "GET",
-		Header: make(http.Header),
-	}
-	w := httptest.NewRecorder()
+	doc := fmt.Sprintf(`hello: Path("/hello") -> setPath("/bye") -> setQuery("void") -> "%s"`, s.URL)
 
-	doc := fmt.Sprintf(`hello: Path("/hello") -> "%s"`, s.URL)
-	tracer := &tracingtest.Tracer{TraceContent: traceContent}
+	tracer := mocktracer.New()
 	params := Params{
 		OpenTracing: &OpenTracingParams{
 			Tracer: tracer,
 		},
 		Flags: FlagsNone,
 	}
 
+	os.Setenv("HOSTNAME", "ingress.tracing.test")
+
 	tp, err := newTestProxyWithParams(doc, params)
 	if err != nil {
-		t.Error(err)
-		return
+		t.Fatal(err)
 	}
 	defer tp.close()
+	if testing.Verbose() {
+		tp.log.Unmute()
+	}
 
-	tp.proxy.ServeHTTP(w, r)
+	ps := httptest.NewServer(tp.proxy)
+	defer ps.Close()
 
-	if len(tracer.RecordedSpans) == 0 {
-		t.Fatal("no span recorded...")
-	}
-	if tracer.RecordedSpans[0].Trace != traceContent {
-		t.Errorf("trace not found, got `%s` instead", tracer.RecordedSpans[0].Trace)
+	req, err := http.NewRequest("GET", ps.URL+"/hello?world", nil)
+	if err != nil {
+		t.Fatal(err)
 	}
+	req.Header.Set("X-Flow-Id", "test-flow-id")
 
-	root, ok := tracer.FindSpan("ingress")
-	if !ok {
-		t.Fatal("root span not found")
+	_, err = http.DefaultClient.Do(req)
+	if err != nil {
+		t.Fatal(err)
 	}
 
-	if len(root.Refs) != 0 {
-		t.Error("root span cannot have references")
-	}
+	// client may get response before proxy finishes span
+	time.Sleep(10 * time.Millisecond)
+
+	span, ok := findSpan(tracer, "ingress")
+	if !ok {
+		t.Fatal("ingress span not found")
+	}
+
+	verifyTag(t, span, SpanKindTag, SpanKindServer)
+	verifyTag(t, span, ComponentTag, "skipper")
+	verifyTag(t, span, HTTPUrlTag, "/hello?world") // For server requests there is no scheme://host:port, see https://golang.org/pkg/net/http/#Request
+	verifyTag(t, span, HTTPMethodTag, "GET")
+	verifyTag(t, span, HostnameTag, "ingress.tracing.test")
+	// verifyTag(t, span, HTTPRemoteAddrTag, "")
+	verifyTag(t, span, HTTPPathTag, "/hello")
+	verifyTag(t, span, HTTPHostTag, ps.Listener.Addr().String())
```

Comment:
I have refactored ingress and proxy span tests to use `httptest.NewServer` instead of calling `ServeHTTP` directly in order to get proper `HTTPHostTag` and `HTTPRemoteAddrTag`. 
Since `tracingtest` was failing the race test, I have refactored tests to use `opentracing-go/mocktracer` instead.